src/api/utils/tradeQuoter.ts

```diff
@@ -0,0 +1,451 @@
+  private isFirmQuote: boolean = true;
+  private slippagePercentage: number = 2;
+  private excludedSources: string[] = ['Kyber', 'Eth2Dai', 'Uniswap', 'Mesh'];
+  private zeroExApiKey: string;
+
+  constructor(zeroExApiKey: string = '') {
+    this.zeroExApiKey = zeroExApiKey;
+  }
+
+  /**
+   * Generates a trade quote for a token pair in a SetToken. This method requires
+   * a token metadata map (passed with the options) which can be generated using
+   * the CoinGeckoDataService in '.api/utils/coingecko.ts'.
+   *
+   * @param  options  QuoteOptions: options / config to generate the quote
+   * @return          TradeQuote: trade quote object
+   */
+  async generate(options: QuoteOptions): Promise<TradeQuote> {
+    const chainId = options.chainId;
+    const feePercentage = options.feePercentage || this.feePercentage;
+    const isFirmQuote = (options.isFirmQuote === false) ? false : this.isFirmQuote;
+    const slippagePercentage = options.slippagePercentage || this.slippagePercentage;
+    const feeRecipient = options.feeRecipient || this.feeRecipient;
+    const excludedSources = options.excludedSources || this.excludedSources;
+
+    const exchangeAdapterName = ZERO_EX_ADAPTER_NAME;
+
+    const {
+      fromTokenAddress,
+      toTokenAddress,
+      fromAddress,
+    } = this.sanitizeAddress(options.fromToken, options.toToken, options.fromAddress);
+
+    const amount = this.sanitizeAmount(options.rawAmount, options.fromTokenDecimals);
+
+    const setOnChainDetails = await options.setToken.fetchSetDetailsAsync(
+      fromAddress, [fromTokenAddress, toTokenAddress]
+    );
+
+    const fromTokenRequestAmount = this.calculateFromTokenAmount(
+      setOnChainDetails,
+      fromTokenAddress,
+      amount
+    );
+
+    const {
+      fromTokenAmount,
+      fromUnits,
+      toTokenAmount,
+      toUnits,
+      calldata,
+    } = await this.fetchZeroExQuote( // fetchQuote (and switch...)
+      fromTokenAddress,
+      toTokenAddress,
+      fromTokenRequestAmount,
+      setOnChainDetails.manager,
+      (setOnChainDetails as any).totalSupply, // Typings incorrect,
+      chainId,
+      isFirmQuote,
+      slippagePercentage,
+      feeRecipient,
+      excludedSources
+    );
+
+    // Sanity check response from quote APIs
+    this.validateQuoteValues(
+      setOnChainDetails,
+      fromTokenAddress,
+      toTokenAddress,
+      fromUnits,
+      toUnits
+    );
+
+    const gas = await this.estimateGasCost(
+      options.tradeModule,
+      fromTokenAddress,
+      fromUnits,
+      toTokenAddress,
+      toUnits,
+      exchangeAdapterName,
+      fromAddress,
+      calldata,
+      setOnChainDetails.manager
+    );
+
+    const coinGecko = new CoinGeckoDataService(chainId);
+    const coinPrices = await coinGecko.fetchCoinPrices({
+      contractAddresses: [this.chainCurrencyAddress(chainId), fromTokenAddress, toTokenAddress],
+      vsCurrencies: [ USD_CURRENCY_CODE, USD_CURRENCY_CODE, USD_CURRENCY_CODE ],
```

Comment:
I think you only need USD currency code once, not `contractAddresses.length()` times